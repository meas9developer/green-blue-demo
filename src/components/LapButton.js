import React from 'react'
import { StyleSheet, Text, TouchableHighlight } from 'react-native'
import PropTypes from 'prop-types'

const styles = StyleSheet.create({
  button: {
    borderWidth: 2,
    height: 100,
    width: 100,
    borderRadius: 50,
    justifyContent: 'center',
    alignItems: 'center'
  }
})

const LapButton = ({onLapButtonPress}) => {
  return (
    <TouchableHighlight
      onPress={onLapButtonPress}
      style={styles.button}
    >
      <Text>Lap button</Text>
    </TouchableHighlight>
  )
}

LapButton.propTypes = {
  onLapButtonPress: PropTypes.func
}
LapButton.defaultProps = {}

export default LapButton

