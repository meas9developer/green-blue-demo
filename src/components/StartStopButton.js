import React from 'react'
import { StyleSheet, Text, TouchableHighlight } from 'react-native'
import PropTypes from 'prop-types'

const styles = StyleSheet.create({
  button: {
    borderWidth: 2,
    height: 100,
    width: 100,
    borderRadius: 50,
    justifyContent: 'center',
    alignItems: 'center'
  },
  startButton: {
    borderColor: 'green'
  }
})

const StartStopButton = ({onStartStopPress, name}) => {
  return (
    <TouchableHighlight
      underlayColor="grey"
      onPress={onStartStopPress}
      style={[styles.button, styles.startButton]}
    >
      <Text>{name}</Text>
    </TouchableHighlight>
  )
}

StartStopButton.propTypes = {
  onStartStopPress: PropTypes.func,
  name: PropTypes.string
}
StartStopButton.defaultProps = {}

export default StartStopButton

