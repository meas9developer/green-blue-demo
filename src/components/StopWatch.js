import React from 'react'
import { StyleSheet, Text, View } from 'react-native'
import StartStopButton from './StartStopButton'
import LapButton from './LapButton'
import LapInfo from './LapInfo'
import formatTime from 'minutes-seconds-milliseconds'

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'stretch'
  },
  header: {
    flex: 1
  },
  counter: {
    flex: 5,
    alignItems: 'center',
    justifyContent: 'center'
  },
  timer: {
    fontSize: 50
  },
  buttons: {
    flex: 3,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-around'
  },
  footer: {
    flex: 1,
    justifyContent: 'flex-start',
    alignItems: 'center'
  }
})

const border = (color) => {
  return {
    borderColor: color,
    borderWidth: 4
  }
}

export default class StopWatch extends React.Component {

  state = {
    timeElapsed: null,
    startStopButtonName: 'Start',
    running: false,
    startTime: null,
    laps: []
  }

  render () {
    return (
      <View style={styles.container}>
        <View style={[styles.header, border('white')]}>
          <View style={[styles.counter, border('white')]}>
            <Text style={styles.timer}>{formatTime(this.state.timeElapsed)}</Text>
          </View>
          <View style={[styles.buttons, border('white')]}>
            <StartStopButton onStartStopPress={this.startTimePressed} name={this.state.startStopButtonName}/>
            <LapButton onLapButtonPress={this.lapButtonPressed}/>
          </View>
        </View>

        <View style={[styles.footer, border('white')]}>
          {this.lapsSection()}
        </View>
      </View>

    )
  }

  startTimePressed = () => {

    // stop timer
    if (this.state.running) {
      clearInterval(this.interval)

      this.setState({
        startStopButtonName: 'Start',
        running: false
      })
      return
    }

    // start timer
    this.setState({
      startStopButtonName: 'Stop',
      startTime: new Date(),
      laps: [],
      running: true
    })

    this.interval = setInterval(() => {
      this.setState({
        timeElapsed: new Date() - this.state.startTime
      })
    }, 30)

  }

  lapButtonPressed = () => {
    const lap = this.state.timeElapsed
    this.setState({
      startTime: new Date(),
      laps: [...this.state.laps, lap]
    })
  }

  lapsSection = () => {
    return this.state.laps.map((time, i) => <LapInfo key={i} lap={time}/>)
  }

}

StopWatch.propTypes = {}
StopWatch.defaultProps = {}



