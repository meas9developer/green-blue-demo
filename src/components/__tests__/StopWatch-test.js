import React from 'react'
import StopWatch from './../StopWatch'
import StartStopButton from './../StartStopButton'
import LapButton from './../LapButton'
import renderer from 'react-test-renderer'
import {shallow} from 'enzyme'

it('renders without crashing', () => {
  const rendered = renderer.create(<StopWatch />).toJSON()
  expect(rendered).toBeTruthy()
  expect(rendered).toMatchSnapshot()
});

 it('Stopwatch has child components', () => {
   const stopWatch = shallow(<StopWatch />)
   expect(shallow(<StopWatch />).find(StartStopButton)).toHaveLength(1)
   expect(shallow(<StopWatch />).find(LapButton)).toHaveLength(1)
 })

describe('Stopwatch buttons', () => {
  it('Stopwatch lap button click', () => {
    const wrapper = shallow(<StopWatch />)
    // console.log(wrapper.state())
    expect(wrapper.find(LapButton)).toHaveLength(1)
    expect(wrapper.state('laps')).toHaveLength(0)
    expect(wrapper.state('startTime')).toBeNull()
    wrapper.find(StartStopButton).simulate('startStopPress')
    wrapper.find(LapButton).simulate('lapButtonPress')
    expect(wrapper.state('laps')).toHaveLength(1)
    expect(wrapper.state('startTime')).not.toBeNull()
    // console.log(wrapper.state())
  })

  it('Stopwatch StartStop button click', () => {
    const wrapper = shallow(<StopWatch />)
    // console.log(wrapper.state())
    expect(wrapper.find(StartStopButton)).toHaveLength(1)
    expect(wrapper.state('startStopButtonName')).toBe('Start')
    expect(wrapper.state('startTime')).toBeNull()
    expect(wrapper.state('running')).toBe(false)
    wrapper.find(StartStopButton).simulate('startStopPress')
    expect(wrapper.state('startStopButtonName')).toBe('Stop')
    expect(wrapper.state('startTime')).not.toBeNull()
    expect(wrapper.state('running')).toBe(true)
    // console.log(wrapper.state())
    wrapper.find(StartStopButton).simulate('startStopPress')
    expect(wrapper.state('startStopButtonName')).toBe('Start')
    expect(wrapper.state('running')).toBe(false)
    // console.log(wrapper.state())
  })

  it('Stopwatch styles testing', () => {
    const wrapper = shallow(<StopWatch />)

    expect(wrapper.props().style.flex).toBe(1)
    expect(wrapper.props().style.alignItems).toBe('stretch')
    expect(wrapper.find('View').at(0).props().style).toEqual({ flex: 1, alignItems: 'stretch' })
    expect(wrapper.find('View').at(1).props().style).toEqual([ { flex: 1 }, { borderColor: 'white', borderWidth: 4 } ])
    expect(wrapper.find('View').at(2).props().style).toEqual([{"alignItems": "center", "flex": 5, "justifyContent": "center"}, {"borderColor": "white", "borderWidth": 4}])
    expect(wrapper.find('View').at(3).props().style).toEqual([ { flex: 3,
      flexDirection: 'row',
      alignItems: 'center',
      justifyContent: 'space-around' },
      { borderColor: 'white', borderWidth: 4 } ])
    expect(wrapper.find('View').at(4).props().style).toEqual([ { flex: 1, justifyContent: 'flex-start', alignItems: 'center' },
      { borderColor: 'white', borderWidth: 4 } ])
  })

})
