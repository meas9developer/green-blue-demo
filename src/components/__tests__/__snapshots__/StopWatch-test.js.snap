// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "stretch",
      "flex": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
        },
        Object {
          "borderColor": "white",
          "borderWidth": 4,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 5,
            "justifyContent": "center",
          },
          Object {
            "borderColor": "white",
            "borderWidth": 4,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "fontSize": 50,
          }
        }
      >
        00:00.00
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flex": 3,
            "flexDirection": "row",
            "justifyContent": "space-around",
          },
          Object {
            "borderColor": "white",
            "borderWidth": 4,
          },
        ]
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 50,
                "borderWidth": 2,
                "height": 100,
                "justifyContent": "center",
                "width": 100,
              },
              Object {
                "borderColor": "green",
              },
            ],
          ]
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          Start
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
            },
            Object {
              "alignItems": "center",
              "borderRadius": 50,
              "borderWidth": 2,
              "height": 100,
              "justifyContent": "center",
              "width": 100,
            },
          ]
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        >
          Lap button
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "flex-start",
        },
        Object {
          "borderColor": "white",
          "borderWidth": 4,
        },
      ]
    }
  />
</View>
`;
